mypy/typeops.py

```diff
@@ -312,46 +312,40 @@ class B(A): pass
             original_type = erase_to_bound(self_param_type)
         original_type = get_proper_type(original_type)
 
-        all_ids = func.type_var_ids()
+        # Find which of method type variables appear in the type of "self".
+        self_ids = {tv.id for tv in get_all_type_vars(self_param_type)}
+        self_vars = [tv for tv in func.variables if tv.id in self_ids]
+
+        # Solve for these type arguments using the actual class or instance type.
         typeargs = infer_type_arguments(
-            func.variables, self_param_type, original_type, is_supertype=True
+            self_vars, self_param_type, original_type, is_supertype=True
         )
         if (
             is_classmethod
-            # TODO: why do we need the extra guards here?
             and any(isinstance(get_proper_type(t), UninhabitedType) for t in typeargs)
             and isinstance(original_type, (Instance, TypeVarType, TupleType))
         ):
-            # In case we call a classmethod through an instance x, fallback to type(x)
+            # In case we call a classmethod through an instance x, fallback to type(x).
             typeargs = infer_type_arguments(
-                func.variables, self_param_type, TypeType(original_type), is_supertype=True
+                self_vars, self_param_type, TypeType(original_type), is_supertype=True
             )
 
-        ids = [tid for tid in all_ids if any(tid == t.id for t in get_type_vars(self_param_type))]
-
+        # Update the method signature with the solutions found.
         # Technically, some constrains might be unsolvable, make them <nothing>.
```

Comment:
Could fix a typo while we're here:

```suggestion
        # Technically, some constraints might be unsolvable, make them <nothing>.
```